### ISSUE TITLE
Spark: Iceberg Data Source does not support struct literal predicates

### ISSUE THREAD

**Issue by szehon-ho** [apache/iceberg#5132]
From the discussion in https://github.com/apache/iceberg/pull/5113 with @huaxingao , I found this behavior:

For Iceberg table:

select * from table where table.struct_field = struct(10)
> org.apache.spark.sql.AnalysisException: cannot resolve '(table.struct_field = struct(10))' due to data type mismatch: differing types in '(table.struct_field = struct(1))' (struct<nested:int> and struct<col1:int>).; line 1 pos 39;

select * from table where table.struct_field in (struct(10))
```
java.lang.IllegalArgumentException: Cannot create expression literal from org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema: [1]
  at org.apache.iceberg.expressions.Literals.from(Literals.java:87)
  at org.apache.iceberg.expressions.UnboundPredicate.<init>(UnboundPredicate.java:40)
  at org.apache.iceberg.expressions.Expressions.equal(Expressions.java:175)
  at org.apache.iceberg.spark.SparkFilters.handleEqual(SparkFilters.java:239)
  at org.apache.iceberg.spark.SparkFilters.convert(SparkFilters.java:152)
  at org.apache.iceberg.spark.source.SparkScanBuilder.pushFilters(SparkScanBuilder.java:106)
  at org.apache.spark.sql.execution.datasources.v2.PushDownUtils$.pushFilters(PushDownUtils.scala:69)
  at org.apache.spark.sql.execution.datasources.v2.V2ScanRelationPushDown$$anonfun$pushDownFilters$1.applyOrElse(V2ScanRelationPushDown.scala:60)
  at org.apache.spark.sql.execution.datasources.v2.V2ScanRelationPushDown$$anonfun$pushDownFilters$1.applyOrElse(V2ScanRelationPushDown.scala:47)
```

For non-Iceberg table:

```
spark.sql("select * from test_struct_non_iceberg where struct_field in(struct(10))").show
+------------+
|struct_field|
+------------+
|        {10}|
+------------+


scala> spark.sql("select * from test_struct_non_iceberg where struct_field = struct(10)").show
+------------+
|struct_field|
+------------+
|        {10}|
+------------+
```

Iceberg cannot handle complex predicate filters (as it does not collect metrics for anything other than primitive columns).  So maybe we should not even push down the filters in SparkScanBuilder.  There may also be other problem (the returned schema not matching)

**Comment by huaxingao** [comment]
You are right. We shouldn't push down the complex predicate filters. I think we should catch the IllegalArgumentException [here](https://github.com/apache/iceberg/blob/master/spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java#L99) and then the complex predicate filter won't be pushed. Similar as [this behavior](https://github.com/apache/iceberg/blob/master/spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java#L106).

I will open a PR to fix this.

**Comment by szehon-ho** [comment]
Thanks @huaxingao , that would be great

**Comment by szehon-ho** [comment]
Actually I had a thought of an optimization, I think for struct column predicate we can pushdown to the nested columns like:

where col= struct{10, 'data'} => col.a = 10 and col.b='data'

It does get a bit trickier with collections (can be IN?)

So the open-ended question is 1) whether we should do it 2) whether Iceberg should do it or Spark should do it generically.  Open to thoughts on this, and doing it as a second phase.

**Comment by sunchao** [comment]
Yes, Spark is already pushing filters on struct fields at the moment:
```
scala> sql("CREATE TABLE foo (s struct<a: int, b: string>) STORED AS PARQUET")
...
scala> sql("SELECT count(*) FROM foo WHERE s.a > 10 AND s.b = 'bar'").explain
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- HashAggregate(keys=[], functions=[count(1)])
   +- Exchange SinglePartition, ENSURE_REQUIREMENTS, [id=#98]
      +- HashAggregate(keys=[], functions=[partial_count(1)])
         +- Project
            +- Filter ((isnotnull(s#95) AND (s#95.a > 10)) AND (s#95.b = bar))
               +- FileScan parquet default.foo[s#95] Batched: false, DataFilters: [isnotnull(s#95), (s#95.a > 10), (s#95.b = bar)], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:..., PartitionFilters: [], PushedFilters: [IsNotNull(s), GreaterThan(s.a,10), EqualTo(s.b,bar)], ReadSchema: struct<s:struct<a:int,b:string>>
```

**Comment by szehon-ho** [comment]
@sunchao oh so I was testing out direct struct literal comparison, like "SELECT count(*) FROM foo WHERE s= struct(10, 'bar')"

**Comment by huaxingao** [comment]
Spark doesn't push down filter for direct struct literal comparison.

**Comment by huaxingao** [comment]
I will work on pushing down struct filter on Spark side. Before the support is in, let's not push down struct filter in iceberg for now. Does this sound OK to you? @szehon-ho

**Comment by szehon-ho** [comment]
Sure @huaxingao , thanks for spending the time on it.

**Comment by szehon-ho** [comment]
Closed by #5204 , we can work on the optimizations later.
